### ISSUE TITLE
Unnecessary null check on local variable after definitely assigning a non-null value

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#39829]
In the collections package, in `lib/src/algorithms.dart`, there is the following method:
```
int binarySearch<T>(List<T> sortedList, T value, {int compare(T a, T b)}) {
  compare ??= defaultCompare<T>();
  int min = 0;
  int max = sortedList.length;
  while (min < max) {
    int mid = min + ((max - min) >> 1);
    var element = sortedList[mid];
    int comp = compare!(element, value);
    if (comp == 0) return mid;
    if (comp < 0) {
      min = mid + 1;
    } else {
      max = mid;
    }
  }
  return -1;
}
```
The parameter `compare` is nullable, but is definitely assigned a non-null value on the first line of the method. The migration engine adds a `!` on the line `int comp = compare!(element, value);`, which seems to be unnecessary.

**Comment by bwilkerson** [comment]
The same thing happens in a slightly different form later in the same file:
```
void reverse(List list, [int start = 0, int? end]) {
  if (end == null) end = list.length;
  _reverse(list, start, end!);
}
```

**Comment by MichaelRFairhurst** [comment]
Refs #40353

**Comment by stereotype441** [comment]
I'm not sure precisely when this was fixed, but as of aa5a5ff5a8d it's no longer occurring.
